lib/xgu/xgu.h

```diff
@@ -0,0 +1,208 @@
+  for(unsigned int i = 0; i < count; i++) {
+    p = push_float(p, f[i]);
+  }
+  return p;
+}
+
+static inline
+DWORD* push_matrix2x2(DWORD* p, float m[2*2]) {
+  return push_floats(p, m, 2*2);
+}
+
+static inline
+DWORD* push_matrix4x4(DWORD* p, float m[4*4]) {
+  return push_floats(p, m, 4*4);
+}
+
+static inline
+DWORD* push_command_matrix2x2(DWORD* p, DWORD command, float m[2*2]) {
+  p = push_command(p, command, 2*2);
+  p = push_matrix2x2(p, m);
+  return p;
+}
+
+static inline
+DWORD* push_command_matrix4x4(DWORD* p, DWORD command, float m[4*4]) {
+  p = push_command(p, command, 4*4);
+  p = push_matrix4x4(p, m);
+  return p;
+}
+
+static inline
+DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+  p = push_command(p, command, 1);
+  p = push_parameter(p, parameter);
+  return p;
+}
+
+inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+  // Force people to use xgu_end instead
+  assert(type != NV097_SET_BEGIN_END_OP_END);
+
+  return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+inline
+DWORD* xgu_end(DWORD* p) {
+  return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_clear_color(DWORD* p, DWORD color) {
+    p = push_command_parameter(p, NV097_SET_COLOR_CLEAR_VALUE, color);
+    return p;
+}
+
+inline
+DWORD* xgu_clear(DWORD* p) {
+    /* Should probably do some magic related to 
+     * NV20_TCL_PRIMITIVE_3D_CLEAR_VALUE_HORIZ and
+     * NV20_TCL_PRIMITIVE_3D_CLEAR_VALUE_DEPTH
+     */
+    return p;
+}
+
+inline
+DWORD* xgu_set_object(DWORD* p, uint32_t instance) {
+    p = push_command(p, NV097_SET_OBJECT, 1);
+    p = push_parameter(p, instance);
+    return p;
+}
+
+inline
+DWORD* xgu_set_model_view_matrix(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_MODEL_VIEW_MATRIX, m);
+}
+
+inline
+DWORD* xgu_matrix_mode_projection(DWORD* p, float m[4*4]) {
+    return push_command_matrix4x4(p, NV097_SET_PROJECTION_MATRIX, m);
+}
+
+inline
+DWORD* xgu_set_texture_matrix(DWORD* p, unsigned int index, float m[2*2]) {
+    assert(index == 0);
+    return push_command_matrix2x2(p, NV097_SET_TEXTURE_MATRIX, m);
+}
+
+inline
+DWORD* xgu_set_viewport(int vpx, int vpy, int vpw, int vph, float znear, float zfar) {
+    
+}
+
+inline
+DWORD* xgu_enable_texture(DWORD* p, int stage_index, bool enabled) {
+    assert(stage_index >= 0);
+    assert(stage_index <= 3);
+    p = push_command(p, NV20_TCL_PRIMITIVE_3D_TX_ENABLE(1), 1);
+    p = push_parameter(p, 0x0003ffc0 | enabled ? (1 << 30) : 0);
+    return p;
+}
+
+#define xgu_set_vertex_array_format(...) 0
+#define xgu_set_vertex_array_offset(...) 0
+
+inline
+DWORD* xgu_elements16(DWORD* p, uint16_t* elements, unsigned int count) {
+  assert(count % 2 == 0);
+  // This one is a bit complicated
+  assert(false);
+  return p;
+}
+
+inline
+DWORD* xgu_elements32(DWORD* p, uint32_t* elements, unsigned int count) {
+  // This one is a bit complicated
+  assert(false);
```

Comment:
**EASY**

See https://github.com/XboxDev/nxdk/blob/master/samples/mesh/main.c#L374-L378

This one isn't too complicated after all.

Reply:
The snippet you linked uses [`NV20_TCL_PRIMITIVE_3D_INDEX_DATA`](https://github.com/XboxDev/nxdk/blob/0605709acd1fa9cf3ae0f652e3ccfa78b248104d/lib/pbkit/nv_objects.h#L1201), defined as `0x00001800` - same as [`NV097_ARRAY_ELEMENT16`](https://github.com/XboxDev/nxdk/blob/0605709acd1fa9cf3ae0f652e3ccfa78b248104d/lib/pbkit/nv_regs.h#L370). Are the defines wrong or is `xgu_elements16` the simple one?

Reply:
`ELEMENT16` and `ELEMENT32` are almost identical, except that `ELEMENT16` always packs 2 indices together per program word (so you have to swap them potentially, making this a more complicated loop). I had probably thought of another issue because this still doesn't make `ELEMENT16` "complicated".

Reply:
I've given it a try to implement these two.